circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,213 @@
+        const auto calculated_hash = hash == 0 ? 0 : preimage.hash();
+        composer.do_assert(hash != calculated_hash,
+                           format("public_call_stack[", i, "] = ", hash, "; does not reconcile"));
+    }
+};
+
+template <typename NT, template <class> typename KernelInput>
+void validate_function_execution(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates state reads and transitions for all type of kernel inputs
+    validate_state_reads(composer, public_kernel_inputs);
+    validate_state_transitions(composer, public_kernel_inputs);
+}
+
+template <typename NT>
+void validate_this_public_call_hash(DummyComposer& composer, PublicKernelInputs<NT> const& public_kernel_inputs)
+{
+    // Pops the current function execution from the stack and validates it against the call stack item
+    const auto& start = public_kernel_inputs.previous_kernel.public_inputs.end;
+    // TODO: this logic might need to change to accommodate the weird edge 3 initial txs (the 'main' tx, the 'fee' tx,
+    // and the 'gas rebate' tx).
+    const auto popped_public_call_hash = array_pop(start.public_call_stack);
+    const auto calculated_this_public_call_hash =
+        public_kernel_inputs.public_call.public_call_data.call_stack_item.hash();
+
+    composer.do_assert(
+        popped_public_call_hash == calculated_this_public_call_hash,
+        "calculated public_call_hash does not match provided public_call_hash at the top of the call stack");
+};
+
+template <typename NT, template <class> typename KernelInput>
+void common_validate_kernel_execution(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates kernel execution for all type of kernel inputs
+    validate_this_public_call_stack(composer, public_kernel_inputs);
+
+    validate_function_execution(composer, public_kernel_inputs);
+}
+
+template <typename NT, template <class> typename KernelInput>
+void common_validate_inputs(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates commons inputs for all type of kernel inputs
+    const auto& this_call_stack_item = public_kernel_inputs.public_call.public_call_data.call_stack_item;
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_delegate_call == false,
+                       "Users cannot make a delegatecall");
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_static_call == false,
+                       "Users cannot make a static call");
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_contract_deployment == false,
+                       "Contract deployment can't be a public function");
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.storage_contract_address ==
+                           this_call_stack_item.contract_address,
+                       "Storage contract address must be that of the called contract");
```

Comment:
Correct only in the case of "no previous kernel". Otherwise, if it's a delegate call, then the storage contract address may differ from the contract address.

Reply:
To be clearer:
- If fn A does a regular call to fn B, then the `storage_contract_address` in fn B's `call_context` is that of fn B. Also, the `msg_sender` in fn B's `call_context` is fn A's contract address.
- If fn A does a delegate call to fn B, then the `storage_contract_address` in fn B's `call_context` is that of fn A. Also, the `msg_sender` in fn B's `call_context` is the same as the `msg_sender` in fn A's `call_context`.